pkg/querier/querier.go

```diff
@@ -191,28 +202,86 @@ func (q querier) Select(sp *storage.SelectParams, matchers ...*labels.Matcher) (
 		}
 	}
 
-	return storage.NewMergeSeriesSet(result, nil), nil, nil
+	// we have all the sets from different sources (chunk from store, chunks from ingesters,
+	// time series from store and time series from ingesters).
+	return q.mergeSeriesSets(result), nil, nil
 }
 
 // LabelsValue implements storage.Querier.
 func (q querier) LabelValues(name string) ([]string, storage.Warnings, error) {
-	lv, err := q.distributor.LabelValuesForLabelName(q.ctx, model.LabelName(name))
-	return lv, nil, err
+	return q.primary.LabelValues(name)
 }
 
 func (q querier) LabelNames() ([]string, storage.Warnings, error) {
-	ln, err := q.distributor.LabelNames(q.ctx)
-	return ln, nil, err
+	return q.primary.LabelNames()
 }
 
-func (q querier) metadataQuery(matchers ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
-	ms, err := q.distributor.MetricsForLabelMatchers(q.ctx, model.Time(q.mint), model.Time(q.maxt), matchers...)
-	if err != nil {
-		return nil, nil, err
+func (querier) Close() error {
+	return nil
+}
+
+func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
+	// Here we deal with sets that are based on chunks and build single set from them.
+	// Remaining sets are merged with chunks-based one using storage.NewMergeSeriesSet
+
+	otherSets := []storage.SeriesSet(nil)
+	chunks := []chunk.Chunk(nil)
+
+	for _, set := range sets {
+		if !set.Next() {
+			// nothing in this set
+			continue
+		}
+
+		s := set.At()
+		if sc, ok := s.(SeriesWithChunks); ok {
+			chunks = append(chunks, sc.Chunks()...)
+
+			// iterate over remaining series in this set, and store chunks
+			// Here we assume that all remaining series in the set are also backed-up by chunks.
+			// If not, there will be panics.
+			for set.Next() {
+				s = set.At()
+				chunks = append(chunks, s.(SeriesWithChunks).Chunks()...)
+			}
+		} else {
+			otherSets = append(otherSets, &ignoreFirstNextSeriesSet{set: set})
+		}
+	}
+
+	if len(chunks) == 0 {
+		return storage.NewMergeSeriesSet(otherSets, nil)
+	}
+
+	// partitionChunks returns set with sorted series, so it can be used by NewMergeSeriesSet
+	chunksSet := partitionChunks(chunks, q.mint, q.maxt, q.chunkIterFn)
+
+	if len(otherSets) == 0 {
+		return chunksSet
 	}
-	return metricsToSeriesSet(ms), nil, nil
+
+	otherSets = append(otherSets, chunksSet)
```

Comment:
The input sets of `NewMergeSeriesSet` must be sorted. How can we guarantee that appending `otherSets ` and `chunksSet` together is still sorted?

Reply:
It's series in individual sets that must be sorted in order for merging to work properly.

Series in chunk-based sets are sorted (guaranteed by `partitionChunks`).

Non-chunk-based sets that we use are:
1. values from ingester when not using streaming – these are returned via `newConcreteSeriesSet`, which sorts the series by label,
2. after 2061 – series set returned by blocks storage, fortunately it also guarantees order of the series.